packages/xeus/src/worker.ts

```diff
@@ -173,25 +193,239 @@ export abstract class EmpackedXeusRemoteKernel extends XeusRemoteKernelBase {
+                updatedInstalledPkgs
+              );
+            }
+          } else {
+            updatedInstalledPkgs = this.deleteFromInstalledPackages(
+              pkgName,
+              updatedInstalledPkgs
+            );
+          }
+        }
+      }
+    });
+
+    return { specs: updatedCurrentSpecs, installed: updatedInstalledPkgs };
+  }
+
+  protected canDelete(
+    specs: string[],
+    installedMap: { [name: string]: ISolvedPackage },
+    type: 'pip' | 'conda'
+  ) {
+    const result: any = [];
+    specs.forEach((spec: string) => {
+      const pkgName = packageNameFromSpec(spec);
+      const specStatus = {
+        status: 1,
+        message: '',
+        spec
+      };
+      if (pkgName) {
+        const pkg = installedMap[pkgName];
+        if (type === 'pip') {
+          if (pkg.repo_name !== 'PyPi') {
+            (specStatus.status = 0),
+              (specStatus.message = `It is imposible to use "pip uninstall" for the package ${spec}, please use "conda remove" command for this`);
+          }
+        } else if (type === 'conda') {
+          if (pkg.repo_name === 'PyPi') {
+            (specStatus.status = 0),
+              (specStatus.message = `It is imposible to use "conda remove" for the package ${spec}, please use "pip uninstall" command for this`);
+          }
+        }
+        result.push(specStatus);
+      }
+    });
+    return result;
+  }
+
+  protected getInstalledPackagesMap() {
+    const newInstalledPackagesMap: { [name: string]: ISolvedPackage } = {};
+    for (const newInstalledPkg of Object.values(this._installedPackages)) {
+      newInstalledPackagesMap[newInstalledPkg.name] = newInstalledPkg;
+    }
+    return newInstalledPackagesMap;
+  }
+
+  protected updateCurrentSpecs(
+    specs: string[],
+    pipSpecs: string[],
+    type: string
+  ): {
+    specs: { [key: string]: string };
+    pipSpecs: { [key: string]: string };
+    installed: ISolvedPackages;
+    status: number;
+  } {
+    // copy data before making manupulations
+    let newSpecs = { ...this._currentSpecs };
+
+    /* we need to collect requested pip specs to know when we need to delete all packages
+       if a user tried to install and after they decided to delete pip packages
+       and if specs array is empty and pip specs array is not, then we should not clean an enviroment
+    */
+    let newPipSpecs = { ...this._currentPipSpecs };
+    let updatedInstalled: ISolvedPackages = { ...this._installedPackages };
+    const newInstalledPackagesMap: { [name: string]: ISolvedPackage } =
+      this.getInstalledPackagesMap();
+
+    let status: number = 1;
+
+    switch (type) {
+      case 'install':
+        newSpecs = this.addSpecs(specs, newSpecs);
+        if (!newInstalledPackagesMap['pip'] && pipSpecs.length) {
+          this.logger.error('Cannot run "pip install": pip is not installed');
+          status = 0;
+          throw new Error('Cannot run "pip install": pip is not installed');
```

Comment:
Thankfully we don't need pip to install pip packages, since we have our own approach. Let's not throw any error and let the user install pip packages.